src/hellogo_fares/read.rs

```diff
@@ -0,0 +1,1123 @@
+// Copyright 2017 Kisio Digital and/or its affiliates.
+//
+// This program is free software: you can redistribute it and/or
+// modify it under the terms of the GNU General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+use super::utils;
+use super::utils::FrameType;
+use super::{TryAttribute, TryOnlyChild};
+use crate::model::Collections;
+use crate::objects::*;
+use crate::Result;
+use failure::{bail, format_err};
+use log::{info, warn};
+use minidom::Element;
+use rust_decimal::Decimal;
+use std::collections::BTreeSet;
+use std::convert::{From, TryFrom};
+use std::fs;
+use std::io::Read;
+use std::path::Path;
+use zip::read::ZipArchive;
+
+impl TryFrom<&Element> for Ticket {
+    type Error = failure::Error;
+    fn try_from(distance_matrix_element: &Element) -> Result<Self> {
+        if distance_matrix_element.name() != "DistanceMatrixElement" {
+            bail!(
+                "Failed to create a ticket from a '{}', it should be a 'DistanceMatrixElement'",
+                distance_matrix_element.name()
+            );
+        }
+        let id = distance_matrix_element.try_attribute("id")?;
+        let ticket = Ticket {
+            id,
+            name: "Ticket Origin-Destination".to_string(),
+            comment: None,
+        };
+        Ok(ticket)
+    }
+}
+
+impl TryFrom<(String, Decimal, String, (Date, Date))> for TicketPrice {
+    type Error = failure::Error;
+    fn try_from(
+        (ticket_id, price, currency, validity): (String, Decimal, String, (Date, Date)),
+    ) -> Result<Self> {
+        iso4217::alpha3(&currency)
+            .ok_or_else(|| format_err!("Failed to convert '{}' as a currency", currency))?;
+        let ticket_price = Self {
+            ticket_id,
+            price,
+            currency,
+            ticket_validity_start: validity.0,
+            ticket_validity_end: validity.1,
+        };
+        Ok(ticket_price)
+    }
+}
+
+impl From<String> for TicketUse {
+    fn from(ticket_id: String) -> Self {
+        let ticket_use_id = format!("TU:{}", ticket_id.clone()).to_string();
```

Comment:
`let ticket_use_id = format!("TU:{}", &ticket_id);`

Reply:
I choose to do it without the `format!()` macro. See 0c7b22f774729f47041525bdbeffc3fed02d392f.